File: src/main/java/net/atos/zac/util/time/ZonedDateTimeParamConverterProvider.java
package net.atos.zac.util.time;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.time.ZonedDateTime;

import jakarta.ws.rs.ext.ParamConverter;
import jakarta.ws.rs.ext.ParamConverterProvider;
import jakarta.ws.rs.ext.Provider;

/**
 * Solves missing valueOf() and string constructor for ZonedDateTime class
 */
@Provider
public class ZonedDateTimeParamConverterProvider implements ParamConverterProvider {

    @SuppressWarnings("unchecked")
    @Override
    public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations) {
        if (rawType.isAssignableFrom(ZonedDateTime.class)) {
            return (ParamConverter<T>) new ZonedDateTimeParamConverter();
        }
        return null;
    }
}

File: src/main/kotlin/nl/info/zac/util/time/ZonedDateTimeParamConverterProvider.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.util.time

import jakarta.ws.rs.ext.ParamConverter
import jakarta.ws.rs.ext.ParamConverterProvider
import jakarta.ws.rs.ext.Provider
import java.lang.reflect.Type
import java.time.ZonedDateTime

/**
 * Provider for ZonedDateTime jakarta web-service parameter converter.
 *
 * @see ZonedDateTimeParamConverter
 */
@Provider
class ZonedDateTimeParamConverterProvider : ParamConverterProvider {
    @Suppress("UNCHECKED_CAST")
    override fun <T> getConverter(rawType: Class<T>, genericType: Type, annotations: Array<Annotation>): ParamConverter<T>? =
        if (rawType.isAssignableFrom(ZonedDateTime::class.java)) {
            ZonedDateTimeParamConverter() as ParamConverter<T>
        } else {
            null
        }
}
